src/tiledb/cloud/vcf/query.py

```diff
@@ -0,0 +1,364 @@
+    :param log_uri: log URI for profiling, defaults to None
+    :param verbose: verbose logging, defaults to False
+    :return: concatenated Arrow table
+    """
+
+    logger = setup(config, verbose)
+
+    with Profiler(array_uri=log_uri) as prof:
+        table = pa.concat_tables(
+            [x for x in tables if x is not None and x.num_rows > 0]
+        )
+        prof.write("result", table.num_rows, table.nbytes)
+
+    memory_usage_gb = max_memory_usage() / (1 << 30)
+    logger.debug("Max memory usage: %0.3f GiB", memory_usage_gb)
+    logger.debug("Records read: %d", table.num_rows)
+    logger.debug("Arrow table size: %0.3f MiB", table.nbytes / (1 << 20))
+
+    return table
+
+
+# --------------------------------------------------------------------
+# DAGs
+# --------------------------------------------------------------------
+
+
+def build_read_dag(
+    dataset_uri: str,
+    *,
+    config: Optional[Mapping[str, Any]] = None,
+    attrs: Optional[Union[Sequence[str], str]] = None,
+    regions: Optional[Union[Sequence[str], str]] = None,
+    bed_file: Optional[str] = None,
+    num_region_partitions: int = 1,
+    max_workers: int = MAX_WORKERS,
+    samples: Optional[Union[Sequence[str], str]] = None,
+    memory_budget_mb: int = 1024,
+    af_filter: Optional[str] = None,
+    log_uri: Optional[str] = None,
+    namespace: Optional[str] = None,
+    resource_class: Optional[str] = None,
+    verbose: bool = False,
+) -> Tuple[tiledb.cloud.dag.DAG, tiledb.cloud.dag.Node]:
+    """
+    Build the DAG for a distributed read on a TileDB-VCF dataset.
+
+    :param dataset_uri: dataset URI
+    :param config: config dictionary, defaults to None
+    :param attrs: attribute names to read, defaults to None
+    :param regions: genomics regions to read, defaults to None
+    :param bed_file: URI of a BED file containing genomics regions to read, defaults to None
+    :param num_region_partitions: number of region partitions, defaults to 1
+    :param samples: sample names to read, defaults to None
+    :param memory_budget_mb: VCF memory budget in MiB, defaults to 1024
+    :param af_filter: allele frequency filter, defaults to None
+    :param log_uri: log array URI for profiling, defaults to None
+    :param namespace: TileDB-Cloud namespace, defaults to None
+    :param resource_class: TileDB-Cloud resource class for UDFs, defaults to None
+    :param verbose: verbose logging, defaults to False
+    :return: DAG and result Node
+    """
+
+    logger = setup(config, verbose)
+
+    # Return an empty table if no samples or regions are specified.
+    # This avoids reading the entire array by accident.
+    if samples is None or (regions is None and bed_file is None):
+        return pa.table({})
+
+    attrs = attrs or DEFAULT_ATTRS
+
+    # Set number of sample partitions
+    num_samples = len(samples)
+    sample_batch_size = ceil(num_samples // (max_workers // num_region_partitions))
```

Comment:
`//` is `__floordiv__`, so `ceil`ing it won’t do anything. Do you want `__truediv__` (i.e., `/`)? (And maybe `num_samples * num_region_partitions / max_workers`?)

Reply:
Right, that math needs some work after some recent changes. Fixing that.